[el8_ppc64le_gcc12] Compilation failure in RecoLocalTracker/SiPixelClusterizer

```
>> Building edm plugin tmp/el8_ppc64le_gcc12/src/RecoLocalTracker/SiPixelClusterizer/plugins/RecoLocalTrackerSiPixelClusterizerPlugins/libRecoLocalTrackerSiPixelClusterizerPlugins.so
.../c++ -O2 -pthread -pipe -Werror=main -Werror=pointer-arith -Werror=overlength-strings -Wno-vla -Werror=overflow -std=c++17 -ftree-vectorize -Werror=array-bounds -Werror=format-contains-nul -Werror=type-limits -fvisibility-inlines-hidden -fno-math-errno --param vect-max-version-for-alias-checks=50 -Xassembler --compress-debug-sections -Wno-error=array-bounds -Warray-bounds -fuse-ld=bfd -fsigned-char -fsigned-bitfields -mcpu=power8 -mtune=power8 --param=l1-cache-size=64 --param=l1-cache-line-size=128 --param=l2-cache-size=512 -felide-constructors -fmessage-length=0 -Wall -Wno-non-template-friend -Wno-long-long -Wreturn-type -Wextra -Wpessimizing-move -Wclass-memaccess -Wno-cast-function-type -Wno-unused-but-set-parameter -Wno-ignored-qualifiers -Wno-deprecated-copy -Wno-unused-parameter -Wunused -Wparentheses -Wno-deprecated -Werror=return-type -Werror=missing-braces -Werror=unused-value -Werror=unused-label -Werror=address -Werror=format -Werror=sign-compare -Werror=write-strings -Werror=delete-non-virtual-dtor -Werror=strict-aliasing -Werror=narrowing -Werror=unused-but-set-variable -Werror=reorder -Werror=unused-variable -Werror=conversion-null -Werror=return-local-addr -Wnon-virtual-dtor -Werror=switch -fdiagnostics-show-option -Wno-unused-local-typedefs -Wno-attributes -Wno-psabi -Wno-error=unused-variable -DBOOST_DISABLE_ASSERTS -shared -Wl,-E -Wl,-z,defs tmp/el8_ppc64le_gcc12/src/RecoLocalTracker/SiPixelClusterizer/plugins/RecoLocalTrackerSiPixelClusterizerPlugins/PixelThresholdClusterizer.cc.o tmp/el8_ppc64le_gcc12/src/RecoLocalTracker/SiPixelClusterizer/plugins/RecoLocalTrackerSiPixelClusterizerPlugins/SiPixelClusterProducer.cc.o tmp/el8_ppc64le_gcc12/src/RecoLocalTracker/SiPixelClusterizer/plugins/RecoLocalTrackerSiPixelClusterizerPlugins/SiPixelDigisClustersFromSoA.cc.o tmp/el8_ppc64le_gcc12/src/RecoLocalTracker/SiPixelClusterizer/plugins/RecoLocalTrackerSiPixelClusterizerPlugins/SiPixelPhase2DigiToClusterCUDA.cc.o tmp/el8_ppc64le_gcc12/src/RecoLocalTracker/SiPixelClusterizer/plugins/RecoLocalTrackerSiPixelClusterizerPlugins/SiPixelRawToClusterCUDA.cc.o tmp/el8_ppc64le_gcc12/src/RecoLocalTracker/SiPixelClusterizer/plugins/RecoLocalTrackerSiPixelClusterizerPlugins/SiPixelRawToClusterGPUKernel.cu.o tmp/el8_ppc64le_gcc12/src/RecoLocalTracker/SiPixelClusterizer/plugins/RecoLocalTrackerSiPixelClusterizerPlugins/RecoLocalTrackerSiPixelClusterizerPlugins_cudadlink.o -o tmp/el8_ppc64le_gcc12/src/RecoLocalTracker/SiPixelClusterizer/plugins/RecoLocalTrackerSiPixelClusterizerPlugins/libRecoLocalTrackerSiPixelClusterizerPlugins.so -Wl,-E -Wl,--hash-style=gnu -L/scratch/cmsbuild/jenkins_a/workspace/build-any-ib/w/tmp/BUILDROOT/6bdb9d190f41eed54da21eaedbdf2bde/opt/cmssw/el8_ppc64le_gcc12/cms/cmssw/CMSSW_13_2_X_2023-06-29-2300/biglib/el8_ppc64le_gcc12 -L/scratch/cmsbuild/jenkins_a/workspace/build-any-ib/w/tmp/BUILDROOT/6bdb9d190f41eed54da21eaedbdf2bde/opt/cmssw/el8_ppc64le_gcc12/cms/cmssw/CMSSW_13_2_X_2023-06-29-2300/lib/el8_ppc64le_gcc12 -L/scratch/cmsbuild/jenkins_a/workspace/build-any-ib/w/tmp/BUILDROOT/6bdb9d190f41eed54da21eaedbdf2bde/opt/cmssw/el8_ppc64le_gcc12/cms/cmssw/CMSSW_13_2_X_2023-06-29-2300/external/el8_ppc64le_gcc12/lib -L/scratch/cmsbuild/jenkins_a/workspace/build-any-ib/w/el8_ppc64le_gcc12/external/cuda/12.0.1-c5280cf8b2a74b3d1a049eae475340f3/lib64/stubs -L/scratch/cmsbuild/jenkins_a/workspace/build-any-ib/w/tmp/BUILDROOT/6bdb9d190f41eed54da21eaedbdf2bde/opt/cmssw/el8_ppc64le_gcc12/cms/cmssw/CMSSW_13_2_X_2023-06-29-2300/static/el8_ppc64le_gcc12 -lRecoTrackerRecord -lCalibTrackerSiPixelESProducers -lEventFilterSiPixelRawToDigi -lTrackingToolsRecords -lCondFormatsSiPixelObjects -lRecoLocalTrackerRecords -lCalibTrackerRecords -lGeometryTrackerGeometryBuilder -lCondFormatsSiStripObjects -lMagneticFieldRecords -lCondFormatsDataRecord -lDataFormatsTrackerCommon -lGeometryCommonTopologies -lCondFormatsAlignment -lDataFormatsGeometryCommonDetAlgo -lDataFormatsSiStripCluster -lGeometryRecords -lGeometryTrackerNumberingBuilder -lCondFormatsAlignmentRecord -lDataFormatsCandidate -lDataFormatsGeometrySurface -lDataFormatsSiPixelDigi -lDataFormatsSiStripCommon -lDataFormatsTrajectoryState -lDetectorDescriptionCore -lDetectorDescriptionDDCMS -lHeterogeneousCoreCUDACore -lCUDADataFormatsSiPixelCluster -lCUDADataFormatsSiPixelDigi -lCondFormatsGeometryObjects -lDataFormatsCLHEP -lDataFormatsEcalDetId -lDataFormatsGeometryVector -lDataFormatsL1GlobalTrigger -lDataFormatsSiPixelDetId -lDataFormatsSiStripDetId -lFWCoreFramework -lHeterogeneousCoreCUDAServices -lCUDADataFormatsCommon -lDataFormatsDetId -lDataFormatsFEDRawData -lDataFormatsL1GlobalMuonTrigger -lDataFormatsMath -lDataFormatsSiPixelCluster -lDataFormatsSiPixelRawData -lDataFormatsSiStripDigi -lFWCoreCommon -lFWCoreServiceRegistry -lCondFormatsPhysicsToolsObjects -lDataFormatsCommon -lFWCoreParameterSet -lHeterogeneousCoreCUDAUtilities -lFWCoreMessageLogger -lDataFormatsProvenance -lFWCorePluginManager -lFWCoreReflection -lCalibFormatsSiPixelObjects -lCondFormatsSerialization -lFWCoreConcurrency -lFWCoreUtilities -lFWCoreVersion -lDDAlign -lDDCond -lDDCore -lDDParsers -lPhysics -lHist -lMatrix -lGenVector -lMathMore -lTree -lNet -lGeom -lThread -lMathCore -lRIO -lboost_iostreams -lboost_serialization -lCore -lboost_thread -lboost_date_time -lCLHEP -lpcre -lbz2 -lcudart -lcudadevrt -lnvToolsExt -lnvidia-ml -lgsl -luuid -ltbb -lxerces-c -llzma -lz -lcuda -lfmt -lcms-md5 -lopenblas -lcrypt -ldl -lrt -lstdc++fs -ltinyxml2
.../ld.bfd: tmp/el8_ppc64le_gcc12/src/RecoLocalTracker/SiPixelClusterizer/plugins/RecoLocalTrackerSiPixelClusterizerPlugins/SiPixelClusterProducer.cc.o: in function `void SiPixelClusterProducer::run<edmNew::DetSetVector<SiPixelCluster> >(edmNew::DetSetVector<SiPixelCluster> const&, edm::ESHandle<TrackerGeometry> const&, edmNew::DetSetVector<SiPixelCluster>&)':
  SiPixelClusterProducer.cc:(.text._ZN22SiPixelClusterProducer3runIN6edmNew12DetSetVectorI14SiPixelClusterEEEEvRKT_RKN3edm8ESHandleI15TrackerGeometryEERS4_[_ZN22SiPixelClusterProducer3runIN6edmNew12DetSetVectorI14SiPixelClusterEEEEvRKT_RKN3edm8ESHandleI15TrackerGeometryEERS4_]+0x1b8): undefined reference to `void PixelThresholdClusterizer::clusterizeDetUnitT<edmNew::DetSet<SiPixelCluster> >(edmNew::DetSet<SiPixelCluster> const&, PixelGeomDetUnit const*, TrackerTopology const*, std::vector<short, std::allocator<short> > const&, edmNew::DetSetVector<SiPixelCluster>::FastFiller&)'
 .../ld.bfd: tmp/el8_ppc64le_gcc12/src/RecoLocalTracker/SiPixelClusterizer/plugins/RecoLocalTrackerSiPixelClusterizerPlugins/SiPixelClusterProducer.cc.o: in function `void SiPixelClusterProducer::run<edm::DetSetVector<PixelDigi> >(edm::DetSetVector<PixelDigi> const&, edm::ESHandle<TrackerGeometry> const&, edmNew::DetSetVector<SiPixelCluster>&)':
  SiPixelClusterProducer.cc:(.text._ZN22SiPixelClusterProducer3runIN3edm12DetSetVectorI9PixelDigiEEEEvRKT_RKNS1_8ESHandleI15TrackerGeometryEERN6edmNew12DetSetVectorI14SiPixelClusterEE[_ZN22SiPixelClusterProducer3runIN3edm12DetSetVectorI9PixelDigiEEEEvRKT_RKNS1_8ESHandleI15TrackerGeometryEERN6edmNew12DetSetVectorI14SiPixelClusterEE]+0x19c): undefined reference to `void PixelThresholdClusterizer::clusterizeDetUnitT<edm::DetSet<PixelDigi> >(edm::DetSet<PixelDigi> const&, PixelGeomDetUnit const*, TrackerTopology const*, std::vector<short, std::allocator<short> > const&, edmNew::DetSetVector<SiPixelCluster>::FastFiller&)'
 .../ld.bfd: tmp/el8_ppc64le_gcc12/src/RecoLocalTracker/SiPixelClusterizer/plugins/RecoLocalTrackerSiPixelClusterizerPlugins/libRecoLocalTrackerSiPixelClusterizerPlugins.so: hidden symbol `_ZN25PixelThresholdClusterizer18clusterizeDetUnitTIN3edm6DetSetI9PixelDigiEEEEvRKT_PK16PixelGeomDetUnitPK15TrackerTopologyRKSt6vectorIsSaIsEERN6edmNew12DetSetVectorI14SiPixelClusterE10FastFillerE' isn't defined
.../ld.bfd: final link failed: bad value
collect2: error: ld returned 1 exit status
  gmake: *** [tmp/el8_ppc64le_gcc12/src/RecoLocalTracker/SiPixelClusterizer/plugins/RecoLocalTrackerSiPixelClusterizerPlugins/libRecoLocalTrackerSiPixelClusterizerPlugins.so] Error 1
```

Full log: [link](https://cmssdt.cern.ch/SDT/cgi-bin/buildlogs/el8_ppc64le_gcc12/CMSSW_13_2_X_2023-06-29-2300/RecoLocalTracker/SiPixelClusterizer)

--- a/RecoLocalTracker/SiPixelClusterizer/plugins/PixelThresholdClusterizer.cc
+++ b/RecoLocalTracker/SiPixelClusterizer/plugins/PixelThresholdClusterizer.cc
@@ -120,80 +120,8 @@ bool PixelThresholdClusterizer::setup(const PixelGeomDetUnit* pixDet) {
 
   return true;
 }
-//----------------------------------------------------------------------------
-//!  \brief Cluster pixels.
-//!  This method operates on a matrix of pixels
-//!  and finds the largest contiguous cluster around
-//!  each seed pixel.
-//!  Input and output data stored in DetSet
-//----------------------------------------------------------------------------
-template <typename T>
-void PixelThresholdClusterizer::clusterizeDetUnitT(const T& input,
-                                                   const PixelGeomDetUnit* pixDet,
-                                                   const TrackerTopology* tTopo,
-                                                   const std::vector<short>& badChannels,
-                                                   edmNew::DetSetVector<SiPixelCluster>::FastFiller& output) {
-  typename T::const_iterator begin = input.begin();
-  typename T::const_iterator end = input.end();
 
-  // this should never happen and the raw2digi does not create empty detsets
-  if (begin == end) {
-    edm::LogError("PixelThresholdClusterizer") << "@SUB=PixelThresholdClusterizer::clusterizeDetUnitT()"
-                                               << " No digis to clusterize";
-  }
-
-  //  Set up the clusterization on this DetId.
-  if (!setup(pixDet))
-    return;
-
-  theDetid = input.detId();
-
-  // Set separate cluster threshold for L1 (needed for phase1)
-  auto clusterThreshold = theClusterThreshold;
-  theLayer = (DetId(theDetid).subdetId() == 1) ? tTopo->pxbLayer(theDetid) : 0;
-  if (theLayer == 1)
-    clusterThreshold = theClusterThreshold_L1;
-
-  //  Copy PixelDigis to the buffer array; select the seed pixels
-  //  on the way, and store them in theSeeds.
-  if (end > begin)
-    copy_to_buffer(begin, end);
-
-  assert(output.empty());
-  //  Loop over all seeds.  TO DO: wouldn't using iterators be faster?
-  for (unsigned int i = 0; i < theSeeds.size(); i++) {
-    // Gavril : The charge of seeds that were already inlcuded in clusters is set to 1 electron
-    // so we don't want to call "make_cluster" for these cases
-    if (theBuffer(theSeeds[i]) >= theSeedThreshold) {  // Is this seed still valid?
-      //  Make a cluster around this seed
-      SiPixelCluster&& cluster = make_cluster(theSeeds[i], output);
-
-      //  Check if the cluster is above threshold
-      // (TO DO: one is signed, other unsigned, gcc warns...)
-      if (cluster.charge() >= clusterThreshold) {
-        // sort by row (x)
-        output.push_back(std::move(cluster));
-        std::push_heap(output.begin(), output.end(), [](SiPixelCluster const& cl1, SiPixelCluster const& cl2) {
-          return cl1.minPixelRow() < cl2.minPixelRow();
-        });
-      }
-    }
-  }
-  // sort by row (x)   maybe sorting the seed would suffice....
-  std::sort_heap(output.begin(), output.end(), [](SiPixelCluster const& cl1, SiPixelCluster const& cl2) {
-    return cl1.minPixelRow() < cl2.minPixelRow();
-  });
-
-  // Erase the seeds.
-  theSeeds.clear();
-
-  //  Need to clean unused pixels from the buffer array.
-  clear_buffer(begin, end);
-
-  theFakePixels.clear();
-
-  thePixelOccurrence.clear();
-}
+#include "PixelThresholdClusterizer.icc"
 
 //----------------------------------------------------------------------------
 //!  \brief Clear the internal buffer array.

--- a/RecoLocalTracker/SiPixelClusterizer/plugins/PixelThresholdClusterizer.icc
+++ b/RecoLocalTracker/SiPixelClusterizer/plugins/PixelThresholdClusterizer.icc
@@ -1,0 +1,74 @@
+//----------------------------------------------------------------------------
+//!  \brief Cluster pixels.
+//!  This method operates on a matrix of pixels
+//!  and finds the largest contiguous cluster around
+//!  each seed pixel.
+//!  Input and output data stored in DetSet
+//----------------------------------------------------------------------------
+template <typename T>
+void PixelThresholdClusterizer::clusterizeDetUnitT(const T& input,
+                                                   const PixelGeomDetUnit* pixDet,
+                                                   const TrackerTopology* tTopo,
+                                                   const std::vector<short>& badChannels,
+                                                   edmNew::DetSetVector<SiPixelCluster>::FastFiller& output) {
+  typename T::const_iterator begin = input.begin();
+  typename T::const_iterator end = input.end();
+
+  // this should never happen and the raw2digi does not create empty detsets
+  if (begin == end) {
+    edm::LogError("PixelThresholdClusterizer") << "@SUB=PixelThresholdClusterizer::clusterizeDetUnitT()"
+                                               << " No digis to clusterize";
+  }
+
+  //  Set up the clusterization on this DetId.
+  if (!setup(pixDet))
+    return;
+
+  theDetid = input.detId();
+
+  // Set separate cluster threshold for L1 (needed for phase1)
+  auto clusterThreshold = theClusterThreshold;
+  theLayer = (DetId(theDetid).subdetId() == 1) ? tTopo->pxbLayer(theDetid) : 0;
+  if (theLayer == 1)
+    clusterThreshold = theClusterThreshold_L1;
+
+  //  Copy PixelDigis to the buffer array; select the seed pixels
+  //  on the way, and store them in theSeeds.
+  if (end > begin)
+    copy_to_buffer(begin, end);
+
+  assert(output.empty());
+  //  Loop over all seeds.  TO DO: wouldn't using iterators be faster?
+  for (unsigned int i = 0; i < theSeeds.size(); i++) {
+    // Gavril : The charge of seeds that were already inlcuded in clusters is set to 1 electron
+    // so we don't want to call "make_cluster" for these cases
+    if (theBuffer(theSeeds[i]) >= theSeedThreshold) {  // Is this seed still valid?
+      //  Make a cluster around this seed
+      SiPixelCluster&& cluster = make_cluster(theSeeds[i], output);
+
+      //  Check if the cluster is above threshold
+      // (TO DO: one is signed, other unsigned, gcc warns...)
+      if (cluster.charge() >= clusterThreshold) {
+        // sort by row (x)
+        output.push_back(std::move(cluster));
+        std::push_heap(output.begin(), output.end(), [](SiPixelCluster const& cl1, SiPixelCluster const& cl2) {
+          return cl1.minPixelRow() < cl2.minPixelRow();
+        });
+      }
+    }
+  }
+  // sort by row (x)   maybe sorting the seed would suffice....
+  std::sort_heap(output.begin(), output.end(), [](SiPixelCluster const& cl1, SiPixelCluster const& cl2) {
+    return cl1.minPixelRow() < cl2.minPixelRow();
+  });
+
+  // Erase the seeds.
+  theSeeds.clear();
+
+  //  Need to clean unused pixels from the buffer array.
+  clear_buffer(begin, end);
+
+  theFakePixels.clear();
+
+  thePixelOccurrence.clear();
+}

--- a/RecoLocalTracker/SiPixelClusterizer/plugins/SiPixelClusterProducer.cc
+++ b/RecoLocalTracker/SiPixelClusterizer/plugins/SiPixelClusterProducer.cc
@@ -199,6 +199,7 @@ void SiPixelClusterProducer::run(const T& input,
   }  // end of DetUnit loop
 }
 
+#include "PixelThresholdClusterizer.icc"
 #include "FWCore/PluginManager/interface/ModuleDef.h"
 #include "FWCore/Framework/interface/MakerMacros.h"
 

